### PR TITLE
[WIP] Brush Interaction for XY-chart

### PR DESCRIPTION
**Pull request by conglei**
💔 Breaking Changes

🏆 Enhancements

Brush interaction is supported! 

more_control_brush

📜 Documentation

🐛 Bug Fix

🏠 Internal

### CODE REVIEW

File: packages/xy-chart/src/index.js [hunk 1]
```diff
@@ -25,3 +25,4 @@ export { PatternLines, PatternCircles, PatternWaves, PatternHexagons } from '@vx
 export { withScreenSize, withParentSize, ParentSize } from '@vx/responsive';
 export { default as withTheme } from './enhancer/withTheme';
 export { chartTheme as theme } from '@data-ui/theme';
+export { default as Brush } from './selection/Brush';
```
Comment (williaster) [inline packages/xy-chart/src/index.js]: do you think all of the brush components should go under `src/components/brush/...` and the utils could go under `src/util/brush`?

File: packages/xy-chart/src/chart/XYChart.jsx [hunk 2]
```diff
@@ -322,6 +323,14 @@ class XYChart extends React.PureComponent {
                 return null;
               } else if (isReferenceLine(name)) {
                 return React.cloneElement(Child, { xScale, yScale });
+              } else if (isBrush(name)) {
+                return React.cloneElement(Child, {
+                  xScale,
+                  yScale,
+                  innerHeight,
```
Comment (williaster) [inline packages/xy-chart/src/chart/XYChart.jsx]: curious if `innerWidth/Height` can be computed from the passed scales?

File: packages/xy-chart/src/utils/brush/Brush.js [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+        end: {
+          x: extent.x1,
+          y: extent.y1,
+        },
+        isBrushing: false,
+      };
+    }));
+  }
+
+  width() {
```
Comment (williaster) [inline packages/xy-chart/src/utils/brush/Brush.js]: nit -- I'd call these `getWidth/Height` for clarity

File: packages/xy-chart/src/utils/brush/Brush.js [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+                  if (onMouseUp) onMouseUp(event);
+                  draw.dragEnd(event);
+                }}
+                style={{ cursor: 'crosshair' }}
+              />
+            );
+          }}
+        </Drag>
+        {/* selection */}
+        {start &&
```
Comment (williaster) [inline packages/xy-chart/src/utils/brush/Brush.js]: if `start`/`end` happened to be `0` these might not work

File: packages/xy-chart/src/utils/brush/Brush.js [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+          end && (
+            <BrushSelection
+              updateBrush={this.update}
+              width={width}
+              height={height}
+              stageWidth={stageWidth}
+              stageHeight={stageHeight}
+              brush={{ ...this.state }}
+              fill="rgba(4, 154, 251, 1.000)"
```
Comment (williaster) [inline packages/xy-chart/src/utils/brush/Brush.js]: could we expose style parameters as `props` do you think?

File: packages/xy-chart/src/utils/brush/Brush.js [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+            const handle = handles[handleKey];
+            return (
+              <BrushHandle
+                key={`handle-${i}`}
+                type={handleKey}
+                handle={handle}
+                stageWidth={stageWidth}
+                stageHeight={stageHeight}
+                updateBrush={this.update}
+                brush={this.state}
```
Comment (williaster) [inline packages/xy-chart/src/utils/brush/Brush.js]: any reason you pass `state` directly here vs copy it for `BrushSelection`?

File: packages/xy-chart/src/utils/brush/RightHandle.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
+import React from 'react';
+import { Drag } from '../drag';
+
+function handleResizeMove(drag, onBrushEnd) {
```
Comment (williaster) [inline packages/xy-chart/src/utils/brush/RightHandle.js]: I'd possibly add these as `static`s

File: packages/xy-chart/src/utils/drag/Drag.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
+import React from 'react';
```
Comment (williaster) [inline packages/xy-chart/src/utils/drag/Drag.js]: I think you'll get some linting errors for `Drag.js` not having a `.jsx` name. Also similar to brush we could move this to a `src/components` folder.

File: packages/xy-chart/src/index.js (hunk 1 above)
Comment (conglei) [reply packages/xy-chart/src/index.js]: So here,  I was thinking for a long term, will we keep the code here or we should move it to vx? The reason I put it in the utils is that I think now it is the temporary solution.

**Comment by conglei** [comment]
@williaster  Now the code is done for the first version. PTLA :)

File: packages/xy-chart/src/chart/XYChart.jsx [hunk 9]
```diff
@@ -349,12 +380,26 @@ class XYChart extends React.PureComponent {
...
             )}
 
+            {hasBrush &&
+              React.cloneElement(Brush, {
+                xScale,
+                yScale,
+                innerHeight,
+                innerWidth,
+                margin,
+                registerStartEvent:
```
[the remark this answers is not in the document]
Comment (conglei) [reply packages/xy-chart/src/chart/XYChart.jsx]: Good catch! In deed for area chart and bar chart, we do need to do like this. Will revise the code logic.

**Comment by conglei** [comment]
@williaster  I've revised the code per comments :) Thanks for catching so many issues :) 

For `react/jsx-handler-names`, as we discussed, all the files with this will be moved to `@vx/brush` and the current naming convention follows the style in `@vx`

**Comment by conglei** [comment]
Hmm, I still haven't figured out why the testings failed here. :(

**Comment by williaster** [comment]
Hmm, I'll try to pull/debug tomorrow, and worse case scenario we can merge to get into DLS. Thanks for making the other fixes!

**Comment by conglei** [comment]
@williaster it seems to be resizing issue. so the bugs happened after the window size changed right ?

**Comment by conglei** [comment]
@williaster  Just fixed the issue. Maybe we can first merge this and I will create an issue for remaining tasks on the brush package.

**Comment by williaster** [comment]
SGTM 👍 Thanks for all of the time spent so far 🙌Excited for this! 🎉

**Pull request merged**